adds getting started cookbook for `ion-schema-rust`

desaikd opened the pull request:
*Description of changes:*
This PR adds a getting started cookbook for `ion-schema-rust`.

*Render view on fork:*
https://desaikd.github.io/ion-schema/docs/cookbook/ion-schema-rust-getting-started

desaikd committed:
adds getting started cookbook for ion-schema-rust

popematt commented on docs/cookbook/ion-schema-rust-getting-started.md:
```diff
@@ -0,0 +1,361 @@
+
+### Creating `DocumentAuthority`:
+
+In general, users will create a [`DocumentAuhtority`](https://docs.rs/ion-schema/0.4.0/ion_schema/authority/trait.DocumentAuthority.html) and then use it to build the [SchemaSystem](https://docs.rs/ion-schema/0.4.0/ion_schema/system/struct.SchemaSystem.html). Then this [SchemaSystem](https://docs.rs/ion-schema/0.4.0/ion_schema/system/struct.SchemaSystem.html) is used instantiating instances of [S ...
```
```suggestion
In general, users will create a [`DocumentAuthority`](https://docs.rs/ion-schema/0.4.0/ion_schema/authority/trait.DocumentAuthority.html) and then use it to build the [SchemaSystem](https://docs.rs/ion-schema/0.4.0/ion_schema/system/struct.SchemaSystem.html). Then this [SchemaSystem](https://docs.rs/ion-schema/0.4.0/ion_schema/system/struct.SchemaSystem.html) is used instantiating instances of [Schema](https://docs.rs/ion-schema/0.4.0/ion_schema/schema/struct.Schema.html).
```

Also, should these links be git blob permalinks instead of links to a particular tag? (I'm not entirely sure myself.)

I actually think that sample code would be more useful than links to the code. The links to the code can quickly go out of date whenever a new version is released, whereas sample code only needs to change as the actual APIs change.

desaikd replied:
I wanted to link the terms used here to rust crate docs. Although it does mean that these links could go out of date. I can remove these links to rust docs for the purpose of this document as this document already has a `Terms` section where all the definitions are defined.

popematt commented on docs/cookbook/ion-schema-rust-getting-started.md:
````diff
@@ -0,0 +1,361 @@
+_Note: A single `SchemaSystem` can contain multiple `DocumentAuthority`s_
+
+```rust
+  // An example of `MapDocumentAuhtority` with `(schema_id, schema content)` pair
````
```suggestion
  // An example of `MapDocumentAuthority` with `(schema_id, schema content)` pair
```

popematt commented on docs/cookbook/ion-schema-rust-getting-started.md:
````diff
@@ -0,0 +1,361 @@
+
+### How to create `MapDocumentAuthority`?
+Creating a `MapDocumentAuthority` requires to pass a `HashMap` with key-value pair of `(schema_id, schema content)`.
+```rust
````
Actually, I think this example makes the example on L54-L69 redundant. I think you should keep this one and get rid of the other one.

popematt commented on docs/cookbook/ion-schema-rust-getting-started.md:
````diff
@@ -0,0 +1,361 @@
+```rust
+// Create authorities vector containing all the authorities that will be used to load a schema based on schema id
+let document_authorities: Vec<Box<dyn DocumentAuthority>> = vec![Box::new(
+    FileSystemDocumentAuthority::new(Path::new("sample_schemas")), <--- provide a path to the authority base folder containing schemas
````
Makes your note into a valid Rust comment:
```suggestion
    FileSystemDocumentAuthority::new(Path::new("sample_schemas")), // <--- provide a path to the authority base folder containing schemas
```

popematt commented on docs/cookbook/ion-schema-rust-getting-started.md:
````diff
@@ -0,0 +1,361 @@
+let mut schema_system = SchemaSystem::new(document_authorities);
+```
+
+## `DocuementAuthority` operations on a `SchemaSystem`
````
```suggestion
## `DocumentAuthority` operations on a `SchemaSystem`
```

popematt commented on docs/cookbook/ion-schema-rust-getting-started.md:
````diff
@@ -0,0 +1,361 @@
+```
+
+### How to add new authority into a `SchemaSystem`?
+
````
Would be worth adding an explanation of _why_ you might want to do this, and where that `authority` goes in terms of priority when resolving SchemaIDs.

popematt replied:
This is what I mean by "where that authority goes in terms of priority when resolving SchemaIDs":
```suggestion
When attempting to resolve a schemaId, `Authority`s are checked in the order in which they were added to the `SchemaSystem`, so any new `Authority` is added with a lower priority than the `Authority`s that are already present.
```

popematt commented on docs/cookbook/ion-schema-rust-getting-started.md:
```diff
@@ -0,0 +1,361 @@
+## `DocuementAuthority` operations on a `SchemaSystem`
+
+### How to get all the authorities that are there for a given `SchemaSystem`?
+
```
Would be worth mentioning _why_ you might want to do this.

popematt commented on docs/cookbook/ion-schema-rust-getting-started.md:
````diff
@@ -0,0 +1,361 @@
+schema_system.add_authority(Box::new(FileSystemDocumentAuthority::new(Path::new("test"))));
+```
+
+### How to create a `SchemsSystem` with given authority or list of authorities?
````
```suggestion
### How to create a `SchemaSystem` with given authority or list of authorities?
```

Or even

```suggestion
### Creating an instance of `SchemaSystem`
```

popematt commented on docs/cookbook/ion-schema-rust-getting-started.md:
````diff
@@ -0,0 +1,361 @@
+
+schema_footer::{
+}
+```
````
This is a repetition of L189-L203. Can you instead have one copy of this sample, and then just refer back to it whenever you need to?

popematt commented on docs/index.md:
```diff
@@ -17,6 +17,7 @@ permalink: /docs/
 * [Modeling SQL Decimals](cookbook/sql-decimals)
 * [Optionally ignoring the occurs requirement for fields](cookbook/ignore-occurs-requirements)
 * [Using Ion Schema Schemas](cookbook/ion-schema-schemas)
+* [Getting started with ion-schema-rust](cookbook/ion-schema-rust-getting-started)
```
Minor formatting suggestion—what do you think? (I'm not entirely sure myself.)
```suggestion
* [Getting started with `ion-schema-rust`](cookbook/ion-schema-rust-getting-started)
```

popematt commented on docs/cookbook.md:
```diff
@@ -7,3 +7,4 @@ title:  Ion Schema Language Cookbook
 * [Modeling Logical Relationships](cookbook/logical-relationships)
 * [Modeling SQL Decimals](cookbook/sql-decimals)
 * [Optionally ignoring the occurs requirement for fields](cookbook/ignore-occurs-requirements.md)
+* [Getting started with ion-schema-rust](cookbook/ion-schema-rust-getting-started)
```
I see that you had to add this in two places. That's not good from a maintainability standpoint. I'm going to figure out some way to get rid of `cookbook.md`.

popematt commented on docs/cookbook/ion-schema-rust-getting-started.md:
```diff
@@ -0,0 +1,361 @@
+# {{ page.title }}
+_(Applies to Ion Schema 1.0.)_
+
+This is a getting started cookbook for `ion-schema-rust`. It includes all the examples for using `ion-schema-rust` like loading a schema, using Ion schema ot validate Ion values etc.
```
```suggestion
This is a getting started cookbook for `ion-schema-rust`. It includes all the examples for using `ion-schema-rust` like loading a schema, using Ion schema to validate Ion values, etc.
```

popematt reviewed (changes requested):
This is great information for a Getting Started guide. I have a few general thoughts, in addition to the specific comments I've made.

* Does "getting started" belong in a cookbook or does it belong in the README of `ion-schema-rust` (or both)? IMO, a good library README has at least a minimal "getting started" section.
* My general approach to cookbook pages is to start with a problem, and explain how to solve it.
  * It's a good idea to explain _why_ you would do something. If you don't have a reason why someone would do X, then you may want to consider leaving X out.
  * A cookbook article doesn't need to be comprehensive—that's what rustdoc is for.
* You may want to consider splitting this into multiple articles. E.g.:
  * _Validating data with a `.isl` file using `ion-schema-rust`_.
  * _Programmatic construction and modification of types using `ion-schema-rust`_.
* Make sure that all of your sample code is valid. If you add commentary/notes into the sample code, make sure they are valid comments for that programming language. 
* In general, I recommend one sentence per line. The markdown engine will convert into regular paragraphs and the user browser will handle any necessary line wrapping. Having one sentence per line makes it easier to read diffs when making changes since you're not shuffling text around to meet some arbitrary column width requirement.

desaikd committed:
adds suggested changes

desaikd committed:
adds minor change in cookbook title

desaikd commented:
>     * Does "getting started" belong in a cookbook or does it belong in the README of `ion-schema-rust` (or both)? IMO, a good library README has at least a minimal "getting started" section.
I was trying to do something similar to the Ion cookbook: https://amzn.github.io/ion-docs/guides/cookbook.html

>     * My general approach to cookbook pages is to start with a problem, and explain how to solve it.
I have added a `How to use this cookbook?` section and also added *why* reasoning for the sections you suggested.

popematt commented on docs/cookbook/ion-schema-rust-getting-started.md:
```diff
@@ -0,0 +1,342 @@
+1. The first step is to load a correct Ion schema. In this step, we use a `SchemaSystem` which verifies the syntax of given schema file for its correctness.
+2. The second step is to use this generated schema from #1 and validate Ion values using the type definitions defined within the schema.
+
+This cookbook starts from examples for #1(including examples for all the required structures to be created to complete #1) and at the end adds examples for #2(How to validate Ion value using `ion-schema-rust`)
```
```suggestion
This cookbook starts from examples for #1 (including examples for all the required structures to be created to complete #1) and at the end adds examples for #2 (How to validate Ion value using `ion-schema-rust`)
```

popematt commented on docs/cookbook/ion-schema-rust-getting-started.md:
```diff
@@ -0,0 +1,342 @@
+* `MapDocumentAuthority`
+    * This authority allows to specify a `HashMap` of schema ids as a key and the schema as value. 
+    * Adding this authority your `SchemaSystem` would allow to resolve all schema ids that are within this authority’s map keys.
+    * In order to load a schema id(e.g. `my_schema`) with this `FileSystemDocumentAuthority` that is added to your `SchemaSystem` would be as simple as: `schema_system.load_schema("my_schema")`
```
```suggestion
    * In order to load a schema id (e.g. `my_schema`) with this `FileSystemDocumentAuthority` that is added to your `SchemaSystem` would be as simple as: `schema_system.load_schema("my_schema")`
```

popematt commented on docs/cookbook/ion-schema-rust-getting-started.md:
```diff
@@ -0,0 +1,342 @@
+* `FileSystemDocumentAuthority` 
+    * This authority allows to specify the base path to where all the schema files are saved (e.g. `/home/USER/schemas/`)
+    * Adding this authority your `SchemaSystem` would allow to resolve all schema ids (relative path to schema file, e.g. `my_schema.isl`) that are within this authority’s base path(e.g. `/home/USER/schemas/`)
+    * In order to load a schema id(e.g. `my_schema.isl`) with this `FileSystemDocumentAuthority` that is added to your `SchemaSystem` would be as simple as: `schema_system.load_schema("my_schema.isl")`
```
```suggestion
    * In order to load a schema id (e.g. `my_schema.isl`) with this `FileSystemDocumentAuthority` that is added to your `SchemaSystem` would be as simple as: `schema_system.load_schema("my_schema.isl")`
```

popematt commented on docs/cookbook/ion-schema-rust-getting-started.md:
```diff
@@ -0,0 +1,342 @@
+
+* `FileSystemDocumentAuthority` 
+    * This authority allows to specify the base path to where all the schema files are saved (e.g. `/home/USER/schemas/`)
+    * Adding this authority your `SchemaSystem` would allow to resolve all schema ids (relative path to schema file, e.g. `my_schema.isl`) that are within this authority’s base path(e.g. `/home/USER/schemas/`)
```
```suggestion
    * Adding this authority your `SchemaSystem` would allow to resolve all schema ids (relative path to schema file, e.g. `my_schema.isl`) that are within this authority’s base path (e.g. `/home/USER/schemas/`)
```

popematt commented on docs/cookbook/ion-schema-rust-getting-started.md:
````diff
@@ -0,0 +1,342 @@
+### How to create `IslSchema`?
+
+```rust
+// The `isl-type` defiend in the previous section cna be used here
````
```suggestion
// The `isl-type` defined in the previous section can be used here
```

popematt commented on docs/cookbook/ion-schema-rust-getting-started.md:
```diff
@@ -0,0 +1,342 @@
+
+_Note: The schema file used in this section is previously defined in `How to load a schema?` section_
+
+### How to validate an Ion value using `Schema`?
```
This is redundant.
```suggestion
```

popematt commented on docs/cookbook/ion-schema-rust-getting-started.md:
````diff
@@ -0,0 +1,342 @@
+
+### Output
+When run, the code above produces the following output:
+```rust
````
Technically, this is output, not Rust.
````suggestion
```
````

popematt reviewed (approved):
There's a bunch of little typos that I've suggested some corrections for. Otherwise, it looks good.

desaikd committed:
adds minor typo changes
